spring-boot-project/spring-boot-tools/spring-boot-gradle-plugin/src/main/java/org/springframework/boot/gradle/tasks/bundling/BootArchiveSupport.java

```diff
@@ -76,7 +76,12 @@
 	void configureManifest(Jar jar, String mainClassName) {
 		Attributes attributes = jar.getManifest().getAttributes();
 		attributes.putIfAbsent("Main-Class", this.loaderMainClass);
-		attributes.putIfAbsent("Start-Class", mainClassName);
+		if (mainClassName == null && attributes.get("Start-Class") == null) {
+			throw new ManifestException("No main class could be resolved");
```

Comment:
I'm not sure that this is the right message for the exception. `BootArchiveSupport` doesn't know anything about the main class being resolved. In fact, neither `BootJar` nor `BootWar` do either. It's only `JavaPluginAction` and `WarPluginAction` that set up `MainClassConvention` for the `mainClassName` property of the default `BootJar` and `BootWar` archives.

It would be interesting to know what the effect would be of changing `MainClassConvention` so that it throws an exception rather than returning `null` when it can't resolve the main class. It would also be interesting to know what the effect would be of removing `@Optional` from `BootArchive.getMainClassName()`.

In summary, I'm not 100% sure of the best way to improve the error message and some experimentation is needed. Would you like to do that and comment with your findings?

Reply:
Yes, I will look into it and try to fix it the way you suggested.

Reply:
@wilkinsona 
Removing `@Optional` causes the bootJarManifestMainClass test to fail with the following exception:

```org.springframework.boot.gradle.docs.PackagingDocumentationTests > bootJarManifestMainClass FAILED
    java.lang.RuntimeException at PackagingDocumentationTests.java:68
        Caused by: org.gradle.testkit.runner.UnexpectedBuildFailure at PackagingDocumentationTests.java:68
```
So it seems it is not finding the main class.

Also, having `MainClassConvention` throw an exception if `resolveMainClass()` fails makes the test fail with the same exception as above. Although, running `gradle build` on a test project without a main method will still produce the correct error message. I will try to get around that and push a solution.

Reply:
Unfortunately, the `Start-Class` property in that particular is added later in the creation of the jar and there is no way  to check if it exists inside `MainClassConvention`, so throwing an exception there causes the jar not to be built, when it should. The only solution I could find that passes all tests and does what it should is the one I suggested, but it is true that `BootArchiveSupport` is not aware of the main class.